jbpm-case-mgmt/jbpm-case-mgmt-impl/src/main/java/org/jbpm/casemgmt/impl/wih/StartCaseWorkItemHandler.java

```diff
@@ -0,0 +1,252 @@
+        if (independent) {
+            Map<String, Object> results = new HashMap<>();
+            results.put(CASE_ID, caseId);
+            try {
+                CaseFileInstance snapshot = caseService.getCaseFileInstance(caseId);
+                
+                results.putAll(snapshot.getData());
+            } catch (CaseNotFoundException e) {
+                // case is already completed
+                logger.debug("Case is already completed, not possible to fetch case file data any more");
+            }
+            logger.debug("Completing directly (without waiting for case instance {} completion) work item with id {}", caseId, workItem.getId()); 
+            manager.completeWorkItem(workItem.getId(), results);
+        } else {
+            // save case id so the abort work item can abort/destroy the case instance
+            ((WorkItemImpl)workItem).setParameter(CASE_ID, caseId);
+            logger.debug("Waiting for case instance {} completion before completing work item with id {}", caseId, workItem.getId());
+        }
+    }
+
+    @Override
+    public void abortWorkItem(WorkItem workItem, WorkItemManager manager) {
+        String caseId = (String) workItem.getParameter(CASE_ID);
+        boolean destroy = true;
+        if (workItem.getParameter(DESTROY_ON_ABORT) != null) {
+            destroy = Boolean.parseBoolean((String) workItem.getParameter(DESTROY_ON_ABORT));
+        }
+                
+        CaseService caseService = (CaseService) ServiceRegistry.get().service(ServiceRegistry.CASE_SERVICE);
+        
+        try {
+            if (destroy) {
+                logger.debug("Case {} is going to be destroyed", caseId);
+                caseService.destroyCase(caseId);
+            } else {
+                logger.debug("Case {} is going to be canceled", caseId);
+                caseService.cancelCase(caseId);
+            }
+        } catch (CaseNotFoundException e) {
+            logger.warn("Case instance {} was not found", caseId);
+        } catch (Exception e) {
+            logger.error("Unexpected error during canceling case instance {}", caseId, e);
+        }
+    }
+    
+    protected CaseFileInstance getCaseFile(KieSession ksession) {
+        if (ksession == null) {
+            logger.debug("No ksession available returning null as case file");
+            return null;
+        }
+        Collection<? extends Object> caseFiles = ksession.getObjects(new ClassObjectFilter(CaseFileInstance.class));
+        if (caseFiles.size() == 0) {
+            logger.debug("No case file available in working memory returning null as case file");
+            return null;
+        }
+        CaseFileInstance caseFile = (CaseFileInstance) caseFiles.iterator().next(); 
+        
+        return caseFile;
+    }
+    
+    protected void parseParameters(WorkItem workItem, Map<String, Object> caseFileData, Map<String, OrganizationalEntity> roleAssignments, Map<String, List<String>> accessRestrictions) {
```

Comment:
Too complex method; every if branch can be delegated into a private method.

Reply:
that would mean iterating over parameters many times so I believe it's better to have more complex method instead of interacting over map 4 times

Reply:
I don't think repeated iterating over map is necessary, I was thinking basically about something in line with:
https://gist.github.com/rsynek/ef94e33f63927854592f99a8b4721a9b

Reply:
Yet another method makes more difficult when debugging but if you strongly believe it will make it better I’ll do it

Reply:
@mswiderski these are suggestions, if you don't believe it will make the code better, no point in changing it. 

My reasoning is that the method combines two level of abstraction - deciding what actions to take based on parameter prefixes and details of these actions. Following principles described in Clean Code (Robert C. Martin), method should deal with only 1 level of abstraction. I think quite the opposite is true about debugging such a method - you add one level to stack trace, truth; on the other hand, the name of the method (if chosen well) serves as a hint of what the method does - making the caller method more obvious and readable.

Whether you change it or not, your call.

Reply:
I choose not to split it into separate methods